### DISCUSSION TITLE
Apache HttpClient 5.2.x

### DISCUSSION THREAD

**Discussion by labkey-adam** [mitre/HTTP-Proxy-Servlet#236] [Show and tell]
No idea if there's interest from the author or other users in migrating to the latest Apache HttpClient (5.2.x), but we've done this migration in our own forked version. PR is here: https://github.com/LabKey/platform/pull/3945/files

Thanks for providing and maintaining this very useful proxy!

Adam

**Comment by dsmiley** [comment]
Hey cool!  I took a brief look.  Do tests pass?
I suppose its worth creating a branch here for this at the least, maybe even releasing it as some 2x alpha

**Comment by labkey-adam** [comment]
I haven't tried running your tests. Unfortunately, our fork has deviated from the official version, with custom extensions, etc., which is why I just pointed at our HC 5.2 version instead of creating a PR against the current version. I'd be willing to help with an official branch, though... releasing it in some form would give me even more incentive to refresh and clean up our forked version.

**Comment by labkey-adam** [comment]
FYI, @dsmiley , we've updated our forked version of `ProxyServlet` to incorporate recent changes, extract most of our extensions, better match your formatting, etc. The diff between ours and yours is now almost entirely the HttpClient 5.2.x changes. You can review our version here: https://github.com/LabKey/platform/blob/develop/api/src/org/mitre/dsmiley/httpproxy/ProxyServlet.java

Let me know if you'd consider taking these changes into your repo.

Thanks,
Adam

**Reply by dsmiley** [reply]
Can you please propose a PR, perhaps Draft, to show?  This makes discussing your proposal more realistic.

I'm guessing 5.x is incompatible with 4.x?  i.e. a user must upgrade.  I'm thinking of doing this for 2.0 which will also offer a Jakarta servlet version.

**Comment by labkey-adam** [comment]
Okay, here's a starting point: https://github.com/mitre/HTTP-Proxy-Servlet/pull/239. I can retarget to a different branch, of course. There are a few changes beyond HttpClient that you may or may not want to accept.

I didn't touch the tests yet because I don't think `LocalTestServer` exists in 5.2.x. Not sure how to proceed there, but wanted to get feedback before going further.

Yes, HttpClient 5.x is incompatible with 4.x. Completely different package names, so the two can co-exist within an application. But `ProxyServlet` needs to use one or the other.

**Comment by dsmiley** [comment]
Could we continue to use LocalTestServer from an older Apache HttpClient?  It isn't pretty but...

**Comment by labkey-adam** [comment]
That could work. I adjusted `ModifyHeadersProxyServlet` (which overrides `ProxyServlet`) and added back a dependency on 4.3.6 for the tests (not sure why the existing "tests" dependency isn't sufficient). I think everything compiles now. We'll see if it all runs...
